core/dbt/clients/_jinja_blocks.py

```diff
@@ -251,21 +301,36 @@ def find_block(self):
         # comments are easy
         if matchgroups.get('comment_start') is not None:
             start = match.start()
+            self.advance(match.end())
             self.expect_comment_end()
             self.blocks.append(BlockData(self.data[start:self.pos]))
             return True
 
-        if matchgroups.get('block_type_name') == 'raw':
+        block_type_name = matchgroups.get('block_type_name')
+
+        if block_type_name == 'raw':
             start = match.start()
             self.expect_raw_end()
             self.blocks.append(BlockData(self.data[start:self.pos]))
             return True
 
-        if matchgroups.get('block_type_name') == 'materialization':
+        if block_type_name == 'materialization':
             self.advance(match.end())
             self.handle_materialization(match)
             return True
 
+        if block_type_name == 'do':
+            # if there is a "block_name" in the match groups, we don't expect a
+            # block as the "block name" is actually part of the do-statement.
+            expect_block = matchgroups.get('block_name') is None
+            self.handle_do(match, expect_block=expect_block)
```

Comment:
seems like you shouldn't need to pass `expect_block` along here since `handle_do` determines whether this is a block or a single tag

Reply:
Oops, good catch - I lost that in one of the many refactors during the development of this PR!

Reply:
Oh wait, no - we need that for the (weird!) case of: `{% do thing %}`

We could remove this by changing the regex code to understand do-tags explicitly, or by changing `handle_do` to backtrack to the last whitespace if there is a `block_name` in the matchgroups, but I don't think either of those ideas are as good as just doing it this way.

I've added a corresponding test case and a more explanatory comment.

Reply:
hmm.... ok, makes sense.

Reply:
One way to think about this: if we don't see a "block name", it still might be a statement - for instance `{% do (x = 1234 + my_value) %}` will fail the regex capture for `block_name` and we won't expect a block from this regex, but then `handle_do` will see that there's more data after the `do` and before the `%}` and react appropriately. But if you do see a "block name", you _must_ be inside a statement and therefore should not expect an `{% enddo %}` tag.